runtime/lua/vim/glob.lua

```diff
@@ -1,93 +1,363 @@
   end
+end
 
-  local function condlist(conds, after)
-    return vim.iter(conds):fold(P(false), function(acc, cond)
-      return acc + cond * after
-    end)
+---@param p vim.lpeg.Pattern Pattern directly after `**/`
+---@return vim.lpeg.Pattern LPeg pattern for `**/p`
+local function dseg(p)
+  return m.P { p + notslash ^ 0 * m.P '/' * m.V(1) }
+end
+
+local g = nil
+
+--- Multiplies conditions for braced expansion (Cartesian product)
+---
+--- @param a string|table First part
+--- @param b string|table Second part
+--- @return string|table Cartesian product of values
+local function mul_cond(a, b)
+  if type(a) == 'string' then
+    if type(b) == 'string' then
+      return a .. b
+    elseif type(b) == 'table' then
+      for i = 1, #b do
+        b[i] = a .. b[i]
+      end
+      return b
+    else
+      return a
+    end
+  elseif type(a) == 'table' then
+    if type(b) == 'string' then
+      for i = 1, #a do
+        a[i] = a[i] .. b
+      end
+      return a
+    elseif type(b) == 'table' then
+      local res = {}
+      local idx = 0
+      for i = 1, #a do
+        for j = 1, #b do
+          idx = idx + 1
+          res[idx] = a[i] .. b[j]
+        end
+      end
+      return res
+    else
+      return a
+    end
+  else
+    return b
   end
+end
 
-  local function mul(acc, m)
-    return acc * m
+--- Combines alternatives in braced patterns
+---
+--- @param a string|table First part
+--- @param b string|table Second part
+--- @return table Combined alternatives
+local function add_cond(a, b)
+  if type(a) == 'string' then
+    if type(b) == 'string' then
+      return { a, b }
+    elseif type(b) == 'table' then
+      table.insert(b, 1, a)
+      return b
+    end
+  elseif type(a) == 'table' then
+    if type(b) == 'string' then
+      table.insert(a, b)
+      return a
+    elseif type(b) == 'table' then
+      for i = 1, #b do
+        table.insert(a, b[i])
+        return a
+      end
+    end
   end
+end
 
-  local function star(stars, after)
-    return (-after * (P(1) - pathsep)) ^ #stars * after
+--- Expands patterns handling segment boundaries
+--- `#` prefix is added for sub-grammar to detect in-segment flag
+---
+---@param a table Array of patterns
+---@param b string Tail string
+---@param inseg boolean Whether inside a path segment
+---@return vim.lpeg.Pattern Expanded pattern
+local function expand(a, b, inseg)
+  for i = 1, #a do
+    if inseg then
+      a[i] = '#' .. a[i]
+    end
+    a[i] = g:match(a[i] .. b)
   end
+  local res = a[1]
+  for i = 2, #a do
+    res = res + a[i]
+  end
+  return res
+end
+
+--- Converts a UTF-8 character to its Unicode codepoint
+---
+---@param utf8_str string UTF-8 character
+---@return number Codepoint value
+local function codepoint(utf8_str)
+  local codepoint = 0
+  local byte_count = 0
+
+  for i = 1, #utf8_str do
+    local byte = utf8_str:byte(i)
+
+    if byte_count ~= 0 then
+      codepoint = bit.bor(bit.lshift(codepoint, 6), bit.band(byte, 0x3F))
```

Comment:
Note that we target PUC Lua, not LuaJIT, so `bit` is not always available. Luckily, _we_ bundle `bit` so you can -- and should -- put `local bit = require('bit')` at the top of this file.